### PR TITLE
Calculate/Persist/Load sparse feature sizes

### PR DESCRIPTION
**Pull request by jupyterjazz**
PR for issue #8750 

**Proposed changes**:
- get sparse feature sizes of current data
- load and store them with the model

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/diet_classifier.py [hunk 1]
```diff
@@ -327,6 +327,7 @@ def __init__(
         entity_tag_specs: Optional[List[EntityTagSpec]] = None,
         model: Optional[RasaModel] = None,
         finetune_mode: bool = False,
+        feature_sizes: Optional[Dict[Text, Dict[Text, List[int]]]] = None,
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:330]: Rename this to `sparse_feature_sizes` to be more accurate.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 2]
```diff
@@ -63,8 +68,20 @@ def featurize_training_examples(
                     attribute, featurizers
                 )
         output.append(attribute_to_features)
-
-    return output
+    # get sparse feature sizes
+    sparse_attributes = []
+    for attr, features in output[0].items():
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:73]: Indexing `output` on `0` will break if `output` is empty list.

Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:73]: Also avoid using shorthand variable names. For example use `attribute` instead of `attr`.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 3]
```diff
@@ -63,8 +68,20 @@ def featurize_training_examples(
...
                 )
         output.append(attribute_to_features)
-
-    return output
+    # get sparse feature sizes
+    sparse_attributes = []
+    for attr, features in output[0].items():
+        if features and features[0].is_sparse():
+            sparse_attributes.append(attr)
+    # we exclude labels at this point
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:76]: Comments should describe the `why` of code, not the `what` of code. `what` is understandable from the code.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 4]
```diff
@@ -63,8 +68,20 @@ def featurize_training_examples(
...
+    if label_attribute in sparse_attributes:
+        sparse_attributes.remove(label_attribute)
+    feature_sizes = {}
+    for attr in sparse_attributes:
+        feature_sizes[attr] = training_examples[0].get_sparse_feature_sizes(
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:81]: Again please be careful when directly indexing over an iterable (`training_examples[0]` can break code for empty lists)

File: rasa/nlu/classifiers/diet_classifier.py [hunk 5]
```diff
@@ -692,9 +694,13 @@ def _create_model_data(
             # no training data are present to train
             return RasaModelData()
 
-        features_for_examples = model_data_utils.featurize_training_examples(
+        (
+            features_for_examples,
+            feature_sizes,
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:699]: Better to rename `feature_sizes` to `sparse_feature_sizes`.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 6]
```diff
@@ -63,8 +68,20 @@ def featurize_training_examples(
...
-
-    return output
+    # get sparse feature sizes
+    sparse_attributes = []
+    for attr, features in output[0].items():
+        if features and features[0].is_sparse():
+            sparse_attributes.append(attr)
+    # we exclude labels at this point
+    if label_attribute in sparse_attributes:
+        sparse_attributes.remove(label_attribute)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:78]: Lets remove the logic of filtering out label attribute outside this function. It adds a weird coupling of two different features for which the function is needed.

File: rasa/nlu/classifiers/diet_classifier.py [hunk 7]
```diff
@@ -839,6 +846,12 @@ def train(
             self.model.compile(
                 optimizer=tf.keras.optimizers.Adam(self.component_config[LEARNING_RATE])
             )
+        else:
+            old_feature_sizes = self._feature_sizes
+            current_feature_sizes = model_data.get_feature_sizes()
+            print(old_feature_sizes, current_feature_sizes)
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:852]:
```suggestion
```

File: rasa/nlu/classifiers/diet_classifier.py [hunk 8]
```diff
@@ -839,6 +846,12 @@ def train(
             self.model.compile(
                 optimizer=tf.keras.optimizers.Adam(self.component_config[LEARNING_RATE])
             )
+        else:
+            old_feature_sizes = self._feature_sizes
+            current_feature_sizes = model_data.get_feature_sizes()
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:851]: Let's add these in the 3rd PR where they are actually needed.

File: rasa/nlu/selectors/response_selector.py [hunk 9]
```diff
@@ -255,6 +255,7 @@ def __init__(
         all_retrieval_intents: Optional[List[Text]] = None,
         responses: Optional[Dict[Text, List[Dict[Text, Any]]]] = None,
         finetune_mode: bool = False,
+        feature_sizes: Optional[Dict[Text, Dict[Text, List[int]]]] = None,
```
Comment (dakshvar22) [inline rasa/nlu/selectors/response_selector.py:258]: Same rename here too.

File: rasa/utils/tensorflow/model_data.py [hunk 10]
```diff
@@ -532,6 +533,31 @@ def add_lengths(
             self.data[key][sub_key].extend([lengths])
             break
 
+    def add_feature_sizes(
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data.py:536]: method name should reflect that only sparse feature sizes are being added.

File: rasa/utils/tensorflow/model_data.py [hunk 11]
```diff
@@ -264,6 +264,7 @@ def __init__(
         self.label_sub_key = label_sub_key
         # should be updated when features are added
         self.num_examples = self.number_of_examples()
+        self.feature_sizes = None
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data.py:267]: Rename to `self.sparse_feature_sizes`

**Review by dakshvar22 — COMMENTED** [review]
Reviewed the code part for now and not the tests.

File: rasa/utils/tensorflow/model_data.py [hunk 12]
```diff
@@ -532,6 +533,31 @@ def add_lengths(
...
+        self, sparse_feature_sizes: Dict[Text, Dict[Text, List[int]]]
+    ) -> None:
+        """Adds a dictionary of feature sizes for different attributes.
+
+        Args:
+            sparse_feature_sizes: a dictionary of attribute that has sparse
+                           features to a dictionary of a feature type
+                           to a list of different sparse feature sizes.
+        """
+        if not self.sparse_feature_sizes:
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data.py:546]: Why do you need the `if` construct?

File: rasa/nlu/selectors/response_selector.py [hunk 13]
```diff
@@ -267,6 +268,7 @@ def __init__(
             responses: All responses defined in the data.
             finetune_mode: If `True` loads the model with pre-trained weights,
                 otherwise initializes it with random weights.
+            sparse_feature_sizes: sizes of the sparse features the model was trained on
```
Comment (dakshvar22) [inline rasa/nlu/selectors/response_selector.py:271]:
```suggestion
            sparse_feature_sizes: Sizes of the sparse features the model was trained on.
```

File: rasa/utils/tensorflow/model_data_utils.py [hunk 14]
```diff
@@ -64,7 +69,21 @@ def featurize_training_examples(
                 )
         output.append(attribute_to_features)
 
-    return output
+    sparse_feature_sizes = {}
+    if output:
+        sparse_attributes = []
+        for attr, features in output[0].items():
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:75]: Please avoid using shorthand variable names like `attr`

File: rasa/utils/tensorflow/model_data_utils.py [hunk 15]
```diff
@@ -64,7 +69,21 @@ def featurize_training_examples(
...
-    return output
+    sparse_feature_sizes = {}
+    if output:
+        sparse_attributes = []
+        for attr, features in output[0].items():
+            if features and features[0].is_sparse():
+                sparse_attributes.append(attr)
+        # we don't support updating label attribute layers at this point
+        if label_attribute in sparse_attributes:
+            sparse_attributes.remove(label_attribute)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:80]: Please address my previous comment - https://github.com/RasaHQ/rasa/pull/8795#discussion_r644078882

File: rasa/utils/tensorflow/model_data_utils.py [hunk 16]
```diff
@@ -64,7 +69,21 @@ def featurize_training_examples(
...
+        if training_examples:
+            for attribute in sparse_attributes:
+                sparse_feature_sizes[attribute] = training_examples[
+                    0
+                ].get_sparse_feature_sizes(attribute=attribute, featurizers=featurizers)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:85]: Please refactor this out in a separate function to keep the original function length short and also allow easy testing of this part of code snippet.

File: tests/shared/nlu/training_data/test_message.py [hunk 17]
```diff
@@ -176,6 +176,100 @@ def test_get_sparse_features(
...
+                ),
+            ],
+            TEXT,
+            ["c1"],
+            [3],
+            None,
+        ),
+    ],
+)
+def test_get_sparse_feature_sizes(
```
Comment (dakshvar22) [inline tests/shared/nlu/training_data/test_message.py:251]: please add test cases where dense features are also part of the features.

File: tests/shared/nlu/training_data/test_message.py [hunk 18]
```diff
@@ -176,6 +176,100 @@ def test_get_sparse_features(
...
+    features: Optional[List[Features]],
+    attribute: Text,
+    featurizers: List[Text],
+    expected_seq_sizes: List[int],
+    expected_sen_sizes: List[int],
```
Comment (dakshvar22) [inline tests/shared/nlu/training_data/test_message.py:256]: Please use `expected_sequence_sizes` and `expected_sentence_sizes`.

File: tests/shared/nlu/training_data/test_message.py [hunk 19]
```diff
@@ -176,6 +176,100 @@ def test_get_sparse_features(
...
+):
+    message = Message(data={TEXT: "This is a test sentence."}, features=features)
+    feature_sizes = message.get_sparse_feature_sizes(attribute, featurizers)
+    if not expected_seq_sizes:
+        assert not feature_sizes[FEATURE_TYPE_SEQUENCE]
```
Comment (dakshvar22) [inline tests/shared/nlu/training_data/test_message.py:261]: Wouldn't `assert feature_sizes[FEATURE_TYPE_SEQUENCE] == expected_seq_sizes` also work even if `expected_seq_sizes` is `None`?

**Review by dakshvar22 — COMMENTED** [review]
Some more comments to be addressed.

File: changelog/8750.improvement.md [hunk 20]
```diff
@@ -0,0 +1 @@
+Changed ``DietClassifier`` class so that we are able to persist and load sparse feature sizes. This will be helpful for incremental training.
```
Comment (wochinge) [inline changelog/8750.improvement.md:1]: is this actually just an internal change? In that case it should be `misc` instead of `improvement`. 
```suggestion
Changed `DietClassifier` class so that we are able to persist and load sparse feature sizes. This will be helpful for incremental training.
```

File: rasa/shared/nlu/training_data/message.py [hunk 21]
```diff
@@ -231,6 +231,38 @@ def get_sparse_features(
...
+            Sparse feature sizes.
+        """
+        if featurizers is None:
+            featurizers = []
+
+        sequence_features, sentence_features = self._filter_sparse_features(
+            attribute, featurizers
+        )
+        sequence_sizes = []
+        if sequence_features:
```
[the remark this answers is not in the document]
Comment (jupyterjazz) [reply rasa/shared/nlu/training_data/message.py:255]: yes, we don't need it. I remember there was a type check error, which I thought I "solved" by adding if statements, but now I realize we don't need them.

File: changelog/8750.improvement.md (hunk 20 above)
Comment (jupyterjazz) [reply changelog/8750.improvement.md:1]: Great 👍  agree it's an internal change and it should be ``misc``.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 22]
```diff
@@ -45,6 +46,8 @@ def featurize_training_examples(
 
     Returns:
         A list of attribute to features.
+        A dictionary of attribute that has sparse features to a dictionary of
+        a feature type to a list of different sparse feature sizes.
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:50]: You don't need to explain the data types of values returned. If you want to explain the schema of the return value, write that in the description of the docstring.

File: rasa/utils/tensorflow/model_data_utils.py [hunk 23]
```diff
@@ -64,7 +67,42 @@ def featurize_training_examples(
...
+    """Collects sparse feature sizes for all attributes that have sparse features.
+
+    Args:
+        featurized_example: a featurized example
+        training_example: a training example
+        featurizers: the featurizers to consider
+
+    Returns:
+        A dictionary of attribute that has sparse features to a dictionary of
+        a feature type to a list of different sparse feature sizes.
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/model_data_utils.py:94]: Same comment here. Explain the schema with an example in the description of the docstring. Examples are way easier to understand for the reader than its description.

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Minor bits to finish up the PR, but looking good 🚀

File: rasa/utils/tensorflow/model_data_utils.py (hunk 3 above)
Comment (dakshvar22) [reply rasa/utils/tensorflow/model_data_utils.py]: Please add comment for the `why` of the code at the place where you have moved this logic to. Currently, the code calls `_remove_label_attribute....` without any explanation of why we do that.

File: tests/utils/tensorflow/test_model_data_utils.py [hunk 24]
```diff
@@ -177,15 +177,31 @@ def test_extract_features():
 
 
 @pytest.mark.parametrize(
-    "text, intent, entities, attributes",
+    "text, intent, entities, attributes, real_feature_sizes",
```
Comment (dakshvar22) [inline tests/utils/tensorflow/test_model_data_utils.py:180]: please be careful with names of the variables. These are `sparse_feature_sizes` to be specific.

File: tests/utils/tensorflow/test_model_data_utils.py [hunk 25]
```diff
@@ -217,8 +234,8 @@ def test_convert_training_examples(
             3,
         )
     ]
-    output = model_data_utils.featurize_training_examples(
-        [message], attributes=attributes, entity_tag_specs=entity_tag_spec
+    output, feature_sizes = model_data_utils.featurize_training_examples(
```
Comment (dakshvar22) [inline tests/utils/tensorflow/test_model_data_utils.py:237]: `feature_sizes` -> `sparse_feature_sizes`

File: rasa/nlu/classifiers/diet_classifier.py [hunk 26]
```diff
@@ -720,6 +730,16 @@ def _create_model_data(
 
         return model_data
 
+    def _remove_label_attribute_feature_sizes(
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:733]: This function can be decorated with `@staticmethod` since it does not use `self` variable anywhere. Pycharm gives you a warning for such functions. Subsequently, please add a test for this function.

Once you make the function static, the test can just call this function as `DIETClassifier._remove_label_attribute_feature_sizes(...)`

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Last few bits that need to be addressed.

File: rasa/nlu/classifiers/diet_classifier.py (hunk 26 above)
Comment (jupyterjazz) [reply rasa/nlu/classifiers/diet_classifier.py:733]: > Once you make the function static, the test can just call this function as DIETClassifier._remove_label_attribute_feature_sizes(...)





Do you mean the ``test_convert_training_examples`` test in ``test_model_data_utils.py`` or the test that will be added in ``test_diet_classifier.py`` in the third PR? 
I think it should go to the second one because we separated the "label removing" logic from ``featurize_training_examples``.

Comment (dakshvar22) [reply rasa/nlu/classifiers/diet_classifier.py:733]: I mean a test for `_remove_label_attribute_feature_sizes` function alone in `test_diet_classifier.py`.

**Review by dakshvar22 — APPROVED** [review]
Thanks for addressing all my comments! Let's first merge https://github.com/RasaHQ/rasa/pull/8771 , then squash and merge this (some tests might fail after you have merged the first PR, please take a look at those).
